src/impl_constructors.rs

```diff
@@ -275,51 +289,52 @@ impl<S, A, D> ArrayBase<S, D>
 
     fn from_shape_vec_impl(shape: StrideShape<D>, v: Vec<A>) -> Result<Self, ShapeError>
     {
+        let dim = shape.dim;
+        let strides = shape.strides;
         if shape.custom {
-            Self::from_vec_dim_stride(shape.dim, shape.strides, v)
+            dimension::can_index_slice(&v, &dim, &strides)?;
         } else {
-            let dim = shape.dim;
-            let strides = shape.strides;
-            if dim.size_checked() != Some(v.len()) {
+            dimension::can_index_slice_not_custom::<A, _>(&v, &dim)?;
+            if dim.size() != v.len() {
                 return Err(error::incompatible_shapes(&Ix1(v.len()), &dim));
             }
-            unsafe { Ok(Self::from_vec_dim_stride_unchecked(dim, strides, v)) }
         }
+        unsafe { Ok(Self::from_vec_dim_stride_unchecked(dim, strides, v)) }
     }
 
-    /// Create an array from a vector and interpret it according to the
-    /// provided dimensions and strides. (No cloning of elements needed.)
+    /// Creates an array from a vector and interpret it according to the
+    /// provided shape and strides. (No cloning of elements needed.)
+    ///
+    /// The caller must ensure that the following conditions are met:
+    ///
+    /// 1. The ndim of `dim` and `strides` must be the same.
+    ///
+    /// 2. The product of non-zero axis lengths must not exceed `isize::MAX`.
+    ///
+    /// 3. For axes with length > 1, the stride must be nonnegative.
```

Comment:
Not sure about why we require this

Reply:
If any axes with length > 1 have negative strides, moving along those axes would result in offsetting the array's pointer backwards outside the `Vec`. (`from_shape_vec_unchecked` is implemented in terms of `from_vec_dim_stride_unchecked`, which [creates the array's pointer from `v.as_mut_ptr()`](https://docs.rs/ndarray/0.12.0/src/ndarray/impl_constructors.rs.html#324).) We could allow negative strides if we allow the user to specify the array's `ptr` (or, equivalently, the offset of `ptr` relative to `v.as_mut_ptr()`).

Reply:
Oh great point, I just never considered that about these constructors. Thanks for the explanation.